### PR TITLE
Use the poll() instead of select() for check state of the socket .

### PR DESCRIPTION
**Pull request by DarovskikhAndrei**
poll used by default. Optionally can use select() (set use_poll=no in configure)
It need for correct working with file descriptors more then 1023.

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -156,6 +156,14 @@ AC_CHECK_HEADERS(netdb.h netinet/in.h sys/socket.h)
 AC_CHECK_HEADERS(sys/param.h sys/select.h inttypes.h)
 AC_CHECK_HEADERS(arpa/inet.h winsock2.h)
 
+use_poll=yes
```
Comment (dinhviethoa) [inline configure.ac]: Could you set poll() not to be used by default for now since it's a risky change?
I'd like to test it for a little bit before making it the default.

Use: AC_ARG_WITH() so that we can use:

```
$ ./configure --with-poll
```

**Comment by DarovskikhAndrei** [comment]
I did it.
select() used by default and new flag --with-poll for use poll()

**Comment by dinhviethoa** [comment]
Could you investigate the build breakage, at least for Linux?

**Comment by DarovskikhAndrei** [comment]
Last commit fix it. No?

File: src/data-types/mailstream_ssl.c [hunk 2]
```diff
@@ -950,17 +998,32 @@ static int wait_write(mailstream_low * s)
...
   FD_SET(ssl_data->fd, &fds_write);
   
-  max_fd = ssl_data->fd;
-  if (fd > max_fd)
-    max_fd = fd;
-  
+  max_fd = fd > ssl_data->fd ? fd : ssl_data->fd;
   r = select(max_fd + 1, &fds_read, &fds_write, NULL, &timeout);
-  if (r <= 0)
+  if (r < 0)
```
Comment (dinhviethoa) [inline src/data-types/mailstream_ssl.c]: why did you change it from <= 0 to < 0?

**Comment by dinhviethoa** [comment]
> Last commit fix it. No?

Thanks! I think when I commented on it, it was still red. It probably happened simultaneously :)

**Comment by DarovskikhAndrei** [comment]
I found more one select() and review fix.

File: src/data-types/mailstream_ssl.c [hunk 3]
```diff
@@ -790,33 +821,45 @@ static int wait_read(mailstream_low * s)
...
   FD_SET(ssl_data->fd, &fds_read);
-  max_fd = ssl_data->fd;
-  if (fd > max_fd)
-    max_fd = fd;
+  max_fd = fd > ssl_data->fd ? fd : ssl_data->fd;
   r = select(max_fd + 1, &fds_read, NULL, NULL, &timeout);
   if (r <= 0)
-    return -1;
-  
-  cancelled = (FD_ISSET(fd, &fds_read));
```
Comment (dinhviethoa) [inline src/data-types/mailstream_ssl.c]: This line disappeared.

File: src/data-types/mailstream_ssl.c [hunk 4]
```diff
@@ -950,17 +998,32 @@ static int wait_write(mailstream_low * s)
...
+  pfd[0].fd = ssl_data->fd;
+  pfd[0].events = POLLOUT;
+  pfd[0].revents = 0;
+
+  pfd[1].fd = fd;
+  pfd[1].events = POLLIN;
+  pfd[1].revents = 0;
+
+  r = poll(&pfd[0], 2, timeout.tv_sec * 1000 + timeout.tv_usec / 1000);
+  if (r < 0)
```
Comment (dinhviethoa) [inline src/data-types/mailstream_ssl.c]: I think you need to test (r <= 0) (similarly to select()) since you want to return an error if there's a timeout.

File: src/data-types/mailstream_ssl.c [hunk 5]
```diff
@@ -790,33 +821,45 @@ static int wait_read(mailstream_low * s)
...
+#elif USE_POLL
+  pfd[0].fd = ssl_data->fd;
+  pfd[0].events = POLLIN;
+  pfd[0].revents = 0;
+
+  pfd[1].fd = fd;
+  pfd[1].events = POLLIN;
+  pfd[1].revents = 0;
+
+  r = poll(&pfd[0], 2, timeout.tv_sec * 1000 + timeout.tv_usec / 1000);
```
Comment (dinhviethoa) [inline src/data-types/mailstream_ssl.c]: same here, if r == 0, there's a timeout error.

File: src/data-types/mailstream_ssl.c (hunk 4 above)
Comment (DarovskikhAndrei) [inline src/data-types/mailstream_ssl.c]: Fix it at all places.

File: src/data-types/mailstream_ssl.c (hunk 5 above)
Comment (DarovskikhAndrei) [inline src/data-types/mailstream_ssl.c]: Fix it at all places.

File: src/data-types/mailstream_ssl.c (hunk 3 above)
Comment (DarovskikhAndrei) [inline src/data-types/mailstream_ssl.c]: Fixed

**Comment by DarovskikhAndrei** [comment]
have you more comments?

**Comment by dinhviethoa** [comment]
It looks all good now! Thanks for helping with that feature!
